quickfixj-core/src/main/java/quickfix/field/converter/IntConverter.java

```diff
@@ -43,20 +46,66 @@ public static String convert(int i) {
      * @see java.lang.Integer#parseInt(String)
      */
     public static int convert(String value) throws FieldConvertError {
-        try {
-            for (int i = 0; i < value.length(); i++) {
-                if (!Character.isDigit(value.charAt(i)) && !(i == 0 && value.charAt(i) == '-')) {
-                    throw new FieldConvertError("invalid integral value: " + value);
+
+        if (!value.isEmpty()) {
+            final char firstChar = value.charAt(0);
+            boolean isNegative = (firstChar == '-');
+            if (!Character.isDigit(firstChar) && !isNegative) {
+                throw new FieldConvertError("invalid integral value: " + value);
+            }
+            int minLength = (isNegative ? 2 : 1);
+            if (value.length() < minLength) {
+                throw new FieldConvertError("invalid integral value: " + value);
+            }
+
+            // Heuristic: since we have no range check in our parseInt() we only parse
+            // values which have at least one digit less than Integer.MAX_VALUE and
+            // leave longer Strings to Integer.parseInt().
+            // NB: we must not simply reject strings longer than MAX_VALUE since
+            // they could possibly include an arbitrary number of leading zeros.
+            int maxLength = (isNegative ? INT_MAX_STRING.length() : INT_MAX_STRING.length() - 1);
+            if (value.length() <= maxLength) {
+                return parseInt(value, isNegative);
+            } else {
+                try {
+                    return Integer.parseInt(value);
+                } catch (NumberFormatException e) {
+                    throw new FieldConvertError("invalid integral value: " + value + ": " + e);
                 }
             }
-            return Integer.parseInt(value);
-        } catch (NumberFormatException e) {
-            throw new FieldConvertError("invalid integral value: " + value + ": " + e);
+        } else {
+            throw new FieldConvertError("invalid integral value: empty string");
+        }
+    }
+
+    /**
+     * Please note that input needs to be validated first, otherwise unexpected
+     * results may occur. Please also note that this method has no range or
+     * overflow check, so please only use it when you are sure that no overflow
+     * might occur (e.g. for parsing seconds or smaller integers).
+     *
+     * This method does however check if the contained characters are digits.
+     *
+     * @param value the String to convert
+     * @param isNegative if passed String is negative, first character will
+     * be skipped since it is assumed that it contains the negative sign
+     * @return the converted int
+     */
+    private static int parseInt(String value, boolean isNegative) throws FieldConvertError {
+        int num = 0;
+        int firstIndex = (isNegative ? 1 : 0);
+        for (int i = firstIndex; i < value.length(); i++) {
+            if (Character.isDigit(value.charAt(i))) {
```

Comment:
Note that [`Character.isDigit(char)`](https://docs.oracle.com/javase/8/docs/api/java/lang/Character.html#isDigit-char-) returns true also for other than the Arabic numerals in the ASCII character range. Would it be better to explicitly check for the range from `'0'` to `'9'`?

Reply:
I used that 0-9 check at first but then realised that formerly already `Character.isDigit(char)` was used in the old code. I checked the implementation and was under the impression that the 0-9 check was used in `Character.isDigit` as well for the Latin charset.
https://github.com/openjdk/jdk/blob/f43bb9feaa03008bad9708a4d7ed850d2532e102/src/java.base/share/classes/java/lang/CharacterDataLatin1.java.template#L84

Reply:
If we already know that the `String` contains only Latin-1 characters, then it will indeed be like that. This logic has been in place since Java 9’s compact strings. In Java 8, [it will still check the Unicode character range](https://github.com/openjdk/jdk/blob/9a9add8825a040565051a09010b29b099c2e7d49/jdk/src/share/classes/java/lang/Character.java#L5668).

Reply:
I didn't check the JDK8 code. So I'll better make it explicit, thank you for your review. :+1: